File: battle_engine/app/contracts/consumers/join_battle_contract.rb
# frozen_string_literal: true

module Contracts
  module Consumers
    class JoinBattleContract < Dry::Validation::Contract
      params do
        required(:player_id).filled(:integer)
        required(:battle_id).filled(:string)
        required(:team_id).filled(:integer)
        optional(:timestamp).maybe(:string)
      end
    end
  end
end

File: battle_real_time/lib/battle_real_time/contracts/publishers/player_actions/join_battle_contract.ex
defmodule BattleRealTime.Contracts.Publishers.PlayerActions.JoinBattleContract do
  @moduledoc """
  Validates the payload for the join_battle player action, under the player_actions queue.
  """
  use BattleRealTime.Contracts.Contract

  @primary_key false
  embedded_schema do
    field(:player_id, :integer)
    field(:battle_id, :string)
    field(:team_id, :integer)
    field(:timestamp, :string)
  end

  def changeset(struct, params) do
    struct
    |> cast(params, [:player_id, :battle_id, :team_id, :timestamp])
    |> validate_required([:player_id, :battle_id, :team_id])
  end
end
